sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,774 @@
+}
+
+func GetItemIDHash(itemID string) string {
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(itemID))
+}
+
+func (s *StoragePackerV2) BucketKeys(ctx context.Context) ([]string, error) {
+	var retErr error
+	s.prewarmCache.Do(func() {
+		diskBuckets, err := logical.CollectKeys(ctx, s.BucketStorageView)
+		if err != nil {
+			retErr = err
+			return
+		}
+		for _, key := range diskBuckets {
+			// Read from the underlying view
+			storageEntry, err := s.BucketStorageView.Get(ctx, key)
+			if err != nil {
+				retErr = errwrap.Wrapf("failed to read packed storage entry: {{err}}", err)
+				return
+			}
+			if storageEntry == nil {
+				retErr = fmt.Errorf("no data found at bucket %s", key)
+				return
+			}
+
+			bucket, err := s.DecodeBucket(storageEntry)
+			if err != nil {
+				retErr = err
+				return
+			}
+
+			s.bucketsCacheLock.Lock()
+			s.bucketsCache.Insert(s.GetCacheKey(bucket.Key), bucket)
+			s.bucketsCacheLock.Unlock()
+		}
+	})
+	if retErr != nil {
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
+
+	return ret, nil
+}
+
+// Get returns a bucket for a given key
+func (s *StoragePackerV2) GetBucket(ctx context.Context, key string, skipCache bool) (*LockedBucket, error) {
+	cacheKey := s.GetCacheKey(key)
+
+	if key == "" {
+		return nil, fmt.Errorf("missing bucket key")
+	}
+
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lock.RLock()
+
+	s.bucketsCacheLock.RLock()
+	_, bucketRaw, found := s.bucketsCache.LongestPrefix(cacheKey)
+	s.bucketsCacheLock.RUnlock()
+
+	if found && !skipCache {
+		ret := bucketRaw.(*LockedBucket)
+		lock.RUnlock()
+		return ret, nil
+	}
+
+	// Swap out for a write lock
+	lock.RUnlock()
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Check for it to have been added
+	s.bucketsCacheLock.RLock()
+	_, bucketRaw, found = s.bucketsCache.LongestPrefix(cacheKey)
+	s.bucketsCacheLock.RUnlock()
+
+	if found && !skipCache {
+		ret := bucketRaw.(*LockedBucket)
+		return ret, nil
+	}
+
+	// Read from the underlying view
+	storageEntry, err := s.BucketStorageView.Get(ctx, key)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to read packed storage entry: {{err}}", err)
+	}
+	if storageEntry == nil {
+		return nil, nil
+	}
+
+	bucket, err := s.DecodeBucket(storageEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	s.bucketsCacheLock.Lock()
+	s.bucketsCache.Insert(cacheKey, bucket)
+	s.bucketsCacheLock.Unlock()
+
+	return bucket, nil
+}
+
+// NOTE: Don't put inserting into the cache here, as that will mess with
+// upgrade cases for the identity store as we want to keep the bucket out of
+// the cache until we actually re-store it.
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
```

Comment:
I was a little concerned that there was no explicit checksum in StoragePackage (storage systems can and do corrupt data), but Snappy does a CRC of its data so we have some coverage there.

Reply:
Because the barrier uses AES-GCM, which is an AEAD, the data is authenticated on decrypt, so we know that it hasn't been modified.

Reply:
Oh, right, there's a decryption function at the next layer down.  Duh.  👍